### ISSUE TITLE
WO-006: planner

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#11]
> **WO-006** - Phase 1 - Owner **MID-strong** - Difficulty **3/5** - Gate **G0**
> Card: [`workorders/WO-006.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-006.md)

## What this is

The planner is the fixed, rule-based counterparty every enterprise faces. This card implements its
six rules: build the planner's view of the world from reported claims (never from the truth), ratchet
next period's targets on reported fulfilment, measure fulfilment, allocate inputs from claimed
availability, physically execute that allocation, and pick which enterprises to audit.

It exists to make the central asymmetry of the whole study concrete: the planner acts on claims. A
claim above stock lowers the delivery pool for the whole good and starves downstream buyers; a claim
below stock leaves the difference sitting in inventory. Padding and shortage are *consequences* of
these six rules, never rules themselves.

## Depends on

- WO-003
- WO-004

The frozen suite this card must pass is WO-002's.

**Blocked by:** #8 (WO-003), #9 (WO-004)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - frozen signatures
  and docstrings of `make_planner_view`, `update_targets`, `fulfilment_measure`, `allocate`,
  `deliver`, `select_audits`, and the `PlannerView` record.
- [`gosplan/env/planner.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/planner.py) -
  **the write target.** Module docstring, the `PlannerView` dataclass and the full contract of all six
  functions are already there; six bodies missing.
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig` and the INFO / INC fields this card reads: `report_lag`, `aggregation_level`,
  `audit_rate`, `audit_mode`, `channel_noise`, `shortfall_visibility`, `quality_measurability`,
  `ratchet_lambda`, `growth_directive`, `ratchet_cap_up`, `ratchet_cap_dn`, `ratchet_deadband`,
  `objective_metric`, `alloc_eta_request`, `alloc_eta_need`, and the TECH field `target_floor_frac`.
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py) - `draw` and
  `PURPOSES`; purposes `audit` and `channel` are this card's.
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py) -
  `State` fields - and the note that only `make_planner_view` and the observation builder may read a
  `State` on someone's behalf.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rule 5 (PLANNER
  BLINDNESS) above all, plus rules 3, 7, 9, 12.
- `docs/ref_worked_example.md` - the hand-validated two-enterprise, two-sector trajectory (written by
  WO-002).
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312.
- [`tests/unit/test_planner.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_planner.py) -
  read-only (CONTRACT rule 2), from WO-002.
- [`tests/behavioural/test_planner_blindness.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/behavioural/test_planner_blindness.py) -
  read-only (CONTRACT rule 2), from WO-002.

Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/env/planner.py`

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/unit/test_planner.py` - **T-U4**: target rule: fixed point at `rho = 1`, `g = 0`; step
      bounded by the caps; floor respected; deadband inert outside `|rho - 1| <= delta`. Plus:
      allocation sums to `avail_j`; `eta_q = 0` ignores requests; `poolfill` in `[0, 1]`; delivery
      conservation.
- [ ] `tests/behavioural/test_planner_blindness.py` - **T-B4**: a state with `S != R` and sentinel
      `y` values produces a `PlannerView` containing no sentinel; and the static check that no
      function in `gosplan/env/planner.py` accepts a `State`.

```bash
uv run pytest tests/unit/test_planner.py -q
uv run pytest tests/behavioural/test_planner_blindness.py -q
uv run ruff check gosplan/env/planner.py
uv run ruff format --check gosplan/env/planner.py
```

## Implementation notes

Six bodies. Every formula is PLAN section 2.7 verbatim; **transcribe, do not re-derive.**

### 1. `make_planner_view(state, cfg) -> PlannerView`

**The only `State -> planner` function in the codebase** (CONTRACT rule 5). It assembles claims,
requests, audit results, measured quality, targets, the planner's I-O estimate, noisy downstream
shortfall (Phase 2), the aggregation level and the plan prices - and nothing else.

The three information filters of PLAN section 2.7.5 are applied here, **in this order**:

```
aggregation    at `aggregation_level = "sector"` the planner sees only
               sum_{i in j} claimed_i, and allocation keys on planned need alone
lag            the rules consume claims from `report_lag` periods ago (Phase 1: 0)
channel noise  claimed_i <- claimed_i * exp(xi_i), xi ~ N(0, sigma_ch**2),
               key = (seed_env, "channel", t, i)
```

**All three branches must exist** even though the Phase-1 configuration makes each of them the
identity; the identity case is what `tests/unit/test_planner.py` checks. The record is built once per
period after the REPORT step, with `audited` all `False` and `audit_meas` all zero, and is reissued
with them filled after `select_audits` and the audit measurement have run.

### 2. `update_targets(view, cfg) -> Array`

PLAN section 2.7.1:

```
m_i    = fulfilment_measure(view, cfg)             # Phase 1: m_i = R_i
rho_i  = m_i / T_i
step_i = clip(rho_i - 1, -c_dn, +c_up)
step_i = 0                       if |rho_i - 1| <= delta      # Phase 1 delta = 0
T_i   <- max(T_min, (1 + g) * T_i * (1 + lambda * step_i))
```

with `T_min = cfg.tech.target_floor_frac * T_0`. **The deadband test is on the RAW `rho_i - 1`**, not
on the clipped step. `g > 0` is the forcing term of finding F1: at `g = 0` with reports at target the
map has a fixed point, which is what T-B2 checks and why `g` is a treatment variable rather than a
constant.

### 3. `fulfilment_measure(view, cfg) -> Array`

PLAN section 2.9.2, three branches:

```
val               m_i = R_i
net_output        m_i = R_i - sum_j p_j * alloc_ij / p_{s(i)}
quality_weighted  m_i = R_i * q_hat_i,   q_hat_i = 1 + mu * (qbar_i - 1)
```

`alloc` is the allocation the planner computed for the SAME view via `allocate(view, cfg)`, `p` is
`view.plan_prices`, `q_hat` is `view.measured_quality`. Every input is planner-side, so this function
never needs the true state. `welfare` is deliberately not an option (finding F6).

### 4. `allocate(view, cfg) -> Array`

PLAN section 2.7.2, promises in **claimed units**:

```
claimed_i = R_i                                      # lagged / noised / aggregated
avail_j   = sum_{i: s(i)=j} (1 - phi_j) * claimed_i
need_bj   = planner_io[s(b), j] * T_b
w_bj      = (q_bj + 1e-6)**eta_q * (need_bj + 1e-6)**eta_n
alloc_bj  = avail_j * w_bj / sum_b w_bj
```

The `1e-6` offsets are part of the formula; **write them, do not adjust them.** At `eta_q = 0` the
request term is exactly 1 and requests are ignored - implement the term anyway, because the test
asserts the Phase-1 behaviour and the Phase-2 value `eta_q = 0.7` is what turns requests into the
hoarding mechanism. `sum_b w_bj` can be zero only if no buyer needs good `j`; in that case
`alloc_bj = 0` for every `b`, and the implementation must produce that without dividing by zero.

### 5. `deliver(state, alloc, cfg) -> (state, deliv, fill, consumer)`

PLAN section 2.7.3:

```
fill_i     = min(1, S_i / claimed_i)        (fill_i = 1 when claimed_i = 0)
shipped_i  = min(S_i, claimed_i)
poolfill_j = sum_{i in j} fill_i * claimed_i / sum_{i in j} claimed_i
deliv_bj   = alloc_bj * poolfill_j
X_bj      += deliv_bj * qbar_j                       # Phase 1 qbar = 1
consumer_j = sum_{i in j} phi_j * shipped_i * qbar_i
S_i       -= shipped_i
```

**Card decision, so that nobody invents one:** when `sum_{i in j} claimed_i == 0` the poolfill
denominator is zero. Every `alloc_bj` for that good is then also zero, so `deliv_bj` is zero for any
finite poolfill; take `poolfill_j = 1.0`, the same convention PLAN section 2.7.3 fixes for `fill_i`
at `claimed_i = 0`, and **say so in your report**. If the frozen test expects something else, file an
AMBIGUITY REPORT and stop.

### 6. `select_audits(view, cfg, t) -> Array`

PLAN section 2.7.4:

```
random    (Phase 1)  audited_i ~ Bernoulli(a),  key = (seed_env, "audit", t, i)
targeted  (Phase 2)  probability a * (1 + kappa_t * downstream_shortfall_i), clipped to
                     [0, 1], active only when shortfall_visibility > 0
```

Phase 1 is `random`. The `targeted` branch's `kappa_t` is **not** a Phase-1 registry parameter; if the
frozen test exercises it, file an AMBIGUITY REPORT rather than inventing a coefficient. The selection
is never observable to any agent before it happens (PLAN section 2.4).

### The `deliver` exception to CONTRACT rule 5, stated plainly

Rule 5 and the T-B4 static check forbid any function in this module except `make_planner_view` from
accepting a `State`. `deliver` takes one. The frozen interface note in `spec/spec.py` resolves it:
`deliver` is the **physical execution** of an allocation already decided from the view - it makes no
planner decision and reads no claim except through `alloc` - so the T-B4 check whitelists it
alongside `make_planner_view`, and the lead records the whitelist (or relocates `deliver` to
`gosplan/env/step.py`) in `spec/CHANGELOG.md` at the v1 freeze.

**Do not relocate it, do not change its signature, and do not add a second `State`-taking function on
the strength of the precedent.** If T-B4 fails on `deliver`, that is a lead and CHANGELOG matter:
file an AMBIGUITY REPORT.

### Rule 7 here

Padding and shortage are CONSEQUENCES of the five lines of `deliver`, never rules. A claim above
stock lowers `poolfill` for the whole good and starves every downstream buyer; a claim below stock
leaves the difference sitting in `S`. CONTRACT rule 7 forbids implementing either directly, and the
lead reviews every `env/` diff against it.

## Forbidden

- Any function in this module taking a `State`, other than `make_planner_view` and the whitelisted
  `deliver`.
- Letting any true quantity cross into `PlannerView`: not `y`, not `S`, not `X`, not `cum_output`,
  not `cum_cost`, not welfare, not the current period's audit selection before it is drawn. T-B4
  plants sentinels in the state and asserts that none of them reaches the view.
- Reading `state.inv_output`, `state.inv_inputs`, `state.cum_output` or `state.cum_cost` anywhere
  except inside `deliver`, where physical execution requires the true stock.
- Any direct `numpy.random` or `jax.random` call (CONTRACT rule 9): `audit` and `channel` draws go
  through `gosplan.rng.draw`.
- Implementing any pathology directly (CONTRACT rule 7): no rule that inflates a request, no rule
  that responds to suspected padding, no bunching-aware allocation.
- Dropping the `lag`, `aggregation` or `channel noise` branch because Phase 1 makes it the identity,
  or reordering the three filters.
- Importing scipy, pandas, jax or torch, at module scope or inside a function.
- Changing any signature, the `PlannerView` fields, or the module docstring; editing any test,
  `spec/spec.py`, or any other module.

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Choosing "the reasonable default" is a violation: do not decide what
  the planner does with information PLAN section 2.7 does not give it. The `targeted` audit
  coefficient and any frozen-test disagreement about `poolfill` are AMBIGUITY REPORTS.
- **Rule 12 - WORK ORDERS (whitelist).** Read only the ten entries above, write only
  `gosplan/env/planner.py`, run the completion command verbatim.
- **Rule 5 - PLANNER BLINDNESS.** Planner rules take a `PlannerView` and nothing else; `PlannerView`
  contains no true quantity. `make_planner_view` is the only builder; `deliver` is the single
  recorded whitelisted exception and must not be generalised.
- **Rule 7 - NO HARD-CODED PATHOLOGY.** No allocation or delivery rule may detect, reward or punish
  padding, hoarding or bunching. The lead reviews the diff before G0.
- **Rule 9 - RNG.** `audit` and `channel` draws go through `gosplan.rng.draw`; no `numpy.random`, no
  module-level generator.
- **Rule 2 - FROZEN TESTS.** Both test files are read-only; a disagreement is an AMBIGUITY REPORT.
- **Rule 1 - FROZEN SPEC.** Signatures, `PlannerView` fields and the module docstring are frozen.

## Definition of done

- [ ] `tests/unit/test_planner.py` (T-U4) green.
- [ ] `tests/behavioural/test_planner_blindness.py` (T-B4) green.
- [ ] `uv run ruff check . && uv run ruff format --check .` clean.
- [ ] `uv run python scripts/contract_guard.py --all` clean.
- [ ] No file outside "Write only" touched.
- [ ] CI `ci-ok` green on the PR.
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output; 2) ambiguity
      reports filed, each listed, or "none"; 3) files changed - expect exactly
      `gosplan/env/planner.py`; 4) anything you were tempted to invent and did not - in particular the
      zero-denominator cases in `allocate` and `deliver`, and anything the `targeted` audit branch
      required.
